src/pyhf/modifiers/shapesys.py

```diff
@@ -66,16 +72,36 @@ def __init__(self, shapesys_mods, pdfconfig, mega_mods, batch_size=None):
             (len(shapesys_mods), self.batch_size or 1, 1),
         )
         # access field is shape (sys, batch, globalbin)
+
+        # reindex it based on current masking
+        self._reindex_access_field(pdfconfig)
+
+        self._precompute()
+        events.subscribe('tensorlib_changed')(self._precompute)
+
+    def _reindex_access_field(self, pdfconfig):
+        # access field maps to bin index, but needs to map to right index in
+        # the parameter set
         for s, syst_access in enumerate(self._access_field):
+            # if the associated shapesys modifier has no parameters (no valid
+            # bins), skip it
+            if not pdfconfig.param_set(self._shapesys_mods[s]).n_parameters:
+                continue
             for t, batch_access in enumerate(syst_access):
                 selection = self.param_viewer.index_selection[s][t]
-                for b, bin_access in enumerate(batch_access):
-                    self._access_field[s, t, b] = (
-                        selection[bin_access] if bin_access < len(selection) else 0
+                access_field_for_syst_and_batch = default_backend.zeros(
+                    len(batch_access)
+                )
+                singular_sample_index = [
+                    i
+                    for i, x in enumerate(
+                        default_backend.astensor(self._shapesys_mask)[s, :, 0]
                     )
-
-        self._precompute()
-        events.subscribe('tensorlib_changed')(self._precompute)
+                    if any(x)
+                ][-1]
+                this_sample_mask = self._shapesys_mask[s][singular_sample_index][0]
```

Comment:
There phrase `sample_mask` doesn't appear anywhere else in the file, so unless I'm missing the importance of the "this" then perhaps just keep it as `sample_mask` to avoid confusion in the future?

```suggestion
                sample_mask = self._shapesys_mask[s][singular_sample_index][0]
```

Reply:
technically, the `this` comes from the fact that it's a loop over samples. So you're saying _this_ sample's mask.